# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [ ]:
import warnings
warnings.simplefilter('ignore')
import pandas as pd
import sqlite3

Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [ ]:
Sales_connection= sqlite3.connect("go_sales.sqlite")
Sales_connection


In [ ]:
sql_query = "SELECT * FROM sqlite_master WHERE type='table';"
df1 = pd.read_sql(sql_query, Sales_connection)
df1

type              name          tbl_name  rootpage  \
0   table           country           country         2   
1   table     order_details     order_details         4   
2   table      order_header      order_header       846   
3   table      order_method      order_method      1072   
4   table           product           product      1074   
5   table      product_line      product_line      1086   
6   table      product_type      product_type      1088   
7   table     retailer_site     retailer_site      1092   
8   table     return_reason     return_reason      1103   
9   table     returned_item     returned_item      1105   
10  table      sales_branch      sales_branch      1133   
11  table       sales_staff       sales_staff      1135   
12  table  SALES_TARGETData  SALES_TARGETData      1141   
13  table   sqlite_sequence   sqlite_sequence      1142   

                                                  sql  
0   CREATE TABLE "country" ("COUNTRY_CODE" VARCHAR...  
1   CREATE TABLE "order_details" ("ORDER_DETAIL_CO...  
2   CREATE TABLE "order_header" ("ORDER_NUMBER" VA...  
3   CREATE TABLE "order_method" ("ORDER_METHOD_COD...  
4   CREATE TABLE "product" ("PRODUCT_NUMBER" VARCH...  
5   CREATE TABLE "product_line" ("PRODUCT_LINE_COD...  
6   CREATE TABLE "product_type" ("PRODUCT_TYPE_COD...  
7   CREATE TABLE "retailer_site" ("RETAILER_SITE_C...  
8   CREATE TABLE "return_reason" ("RETURN_REASON_C...  
9   CREATE TABLE "returned_item" ("RETURN_CODE" VA...  
10  CREATE TABLE "sales_branch" ("SALES_BRANCH_COD...  
11  CREATE TABLE "sales_staff" ("SALES_STAFF_CODE"...  
12  CREATE TABLE "SALES_TARGETData" ("Id" INTEGER ...  
13             CREATE TABLE sqlite_sequence(name,seq)

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [ ]:


newQuery = "SELECT * FROM product"
product = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM product_type"
product_type = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM product_line"
product_line = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM sales_staff"
sales_staff = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM sales_branch"
sales_branch = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM retailer_site"
retailer_site = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM country"
country = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM order_header"
order_header = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM order_details"
order_details = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM SALES_TARGETData"
target = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM returned_item"
returned_item = pd.read_sql(newQuery,Sales_connection)
newQuery = "SELECT * FROM return_reason"
return_reason = pd.read_sql(newQuery,Sales_connection)


Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [ ]:
productkosten = product['PRODUCTION_COST'].astype(float)
productiekosten_tussen_50_en_100 = (productkosten < 100) & (productkosten > 50)

product.loc[productiekosten_tussen_50_en_100 ,['PRODUCT_NUMBER','PRODUCTION_COST']]

PRODUCT_NUMBER PRODUCTION_COST
13            110              60
14            111            81.8
17            114              80
24             17              60
25             18              86
28             20           78.55
31             23           65.33
32             24            62.5
56             46           52.86
58             48              55
69             58           57.13
71              6           85.11
81             69           73.33
83             70              80
84             71              60
85             72            72.5
88             75              80
91             78           79.19
92             79           92.58
93              8              75
94             80           94.12
97             83           56.25
98             84           78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [ ]:
product_margin = product['MARGIN'].astype(float)
product_margin_lager_dan_20_of_hoger_dan_60 = (product_margin < 0.20) | (product_margin > 0.6)
product.loc[product_margin_lager_dan_20_of_hoger_dan_60,['PRODUCT_NUMBER','MARGIN']]

PRODUCT_NUMBER MARGIN
29              21    .17
32              24    .17
100             86    .67
101             87    .67
102             88     .7
103             89    .63
105             90    .63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [ ]:
country_courency_is_fracs = country['CURRENCY_NAME'] == 'francs'
country.loc[country_courency_is_fracs, 'COUNTRY']

0          France
7         Belgium
19    Switzerland
Name: COUNTRY, dtype: object

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [ ]:
margin_hoger_dan_50 = product_margin > 0.50
product.loc[margin_hoger_dan_50, ['INTRODUCTION_DATE']].drop_duplicates('INTRODUCTION_DATE')

INTRODUCTION_DATE
4          10-12-2019
6          18-12-2019
7          27-12-2019
9           13-1-2020
14         15-12-2019
79         26-10-2016
100         15-2-2011

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 7 rijen)

In [ ]:
heeft_tweede_address = sales_branch['ADDRESS2'].notna()
heeft_regio = sales_branch['REGION'].notna()
sales_branch.loc[heeft_tweede_address & heeft_regio,['ADDRESS1','CITY']]

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [ ]:
naam_bevat_dollar = country['CURRENCY_NAME'].str.contains('dollar')
country.loc[naam_bevat_dollar,['COUNTRY']].sort_values(by=['COUNTRY'])

COUNTRY
6       Australia
15         Canada
3          Taiwan
14  United States

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [ ]:
returned_item['RETURN_QUANTITY'].astype(float).sum()

14306.0

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [ ]:
sales_branch.loc[:,['REGION']].drop_duplicates('REGION').count()

REGION    15
dtype: int64

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [ ]:
import numpy as np
product_margin_min = product_margin.min()
product_margin_max = product_margin.max()
product_margin_avg = product_margin.sum() / product_margin.count()
np.mean(product_margin)
df_empty = pd.DataFrame([])
df_empty.agg(product_margin_max,np.mean(product_margin))

ValueError: No axis named 0.396608695652174 for object type DataFrame

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [ ]:
adress_is_null = retailer_site['ADDRESS2'].isna()
retailer_site.loc[adress_is_null,['ADDRESS2']].isna().count()

ADDRESS2    286
dtype: int64

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [ ]:
unit_price = order_details['UNIT_PRICE'].astype(float)
unit_sale_price = order_details['UNIT_SALE_PRICE'].astype(float)

unit_sale_price_lager_dan_unit_price = unit_sale_price < unit_price

unit_cost = order_details.loc[unit_sale_price_lager_dan_unit_price, ['UNIT_COST','QUANTITY']].astype(float)
unit_cost_avg = (unit_cost['UNIT_COST'] * unit_cost['QUANTITY']).sum() / unit_cost['QUANTITY'].sum()
unit_cost_avg 


77.68605100642756

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [ ]:
sales_staff.groupby('POSITION_EN', as_index= False)['SALES_STAFF_CODE'].count()

POSITION_EN  SALES_STAFF_CODE
0                Branch Manager                19
1              District Manager                 4
2               General Manager                 1
3  Level 1 Sales Representative                13
4  Level 2 Sales Representative                39
5  Level 3 Sales Representative                24
6              Regional Manager                 2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [ ]:
sales_staff_work_phone = sales_staff.groupby('WORK_PHONE',as_index=False)['SALES_STAFF_CODE'].count()
sales_staff_work_phone_4 = sales_staff_work_phone['SALES_STAFF_CODE'] > 4
sales_staff_work_phone.loc[sales_staff_work_phone_4, ['WORK_PHONE','SALES_STAFF_CODE']]

WORK_PHONE  SALES_STAFF_CODE
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6

## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [ ]:
retailer_country_merge = pd.merge(retailer_site, country,left_on='COUNTRY_CODE',how='inner',right_on='COUNTRY_CODE')
uit_Nederland = retailer_country_merge['COUNTRY'] == 'Netherlands'
retailer_country_merge.loc[uit_Nederland,['ADDRESS1','CITY']]


ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [ ]:
product_merge = pd.merge(product, product_type, left_on='PRODUCT_TYPE_CODE',how='inner',right_on='PRODUCT_TYPE_CODE')
product_merge.loc[product_merge['PRODUCT_TYPE_EN']=='Eyewear',['PRODUCT_NAME']]

PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [ ]:
sales_staff_merge = pd.merge(sales_staff,order_header,left_on='SALES_STAFF_CODE',how='left',right_on='SALES_STAFF_CODE')
retailer_site_merge = pd.merge(sales_staff_merge, retailer_site, left_on='RETAILER_SITE_CODE',how='left',right_on='RETAILER_SITE_CODE')
retailer_site_merge.loc[(retailer_site_merge['POSITION_EN']=='Branch Manager')&(retailer_site_merge['ORDER_DATE'].notna()),['FIRST_NAME','LAST_NAME','ORDER_DATE']].drop_duplicates('FIRST_NAME')

FIRST_NAME LAST_NAME  ORDER_DATE
5325     Bayard     Lopes  2022-09-09

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [ ]:
sales_staff_merge = pd.merge(sales_staff,order_header,left_on='SALES_STAFF_CODE',how='left',right_on='SALES_STAFF_CODE')
is_manager = sales_staff_merge['POSITION_EN'].str.contains('Manager')
sales_staff_merge.loc[sales_staff_merge['POSITION_EN'].str.contains('Manager'),['SALES_STAFF_CODE','POSITION_EN','ORDER_DATE']].drop_duplicates(['SALES_STAFF_CODE','POSITION_EN']).drop_duplicates()


SALES_STAFF_CODE       POSITION_EN  ORDER_DATE
421               107    Branch Manager         NaN
422               108   General Manager         NaN
607               111    Branch Manager         NaN
881               119    Branch Manager         NaN
1276               13  Regional Manager         NaN
1277               14    Branch Manager         NaN
1642               17    Branch Manager         NaN
1643               18  District Manager         NaN
1730               20    Branch Manager         NaN
1975               25    Branch Manager         NaN
2381               30    Branch Manager         NaN
2382                4    Branch Manager         NaN
2479               45    Branch Manager         NaN
2507               47  District Manager         NaN
2508               48    Branch Manager         NaN
2933               53    Branch Manager         NaN
3351               57    Branch Manager         NaN
3723               61    Branch Manager         NaN
3881               65    Branch Manager         NaN
4347               71  District Manager         NaN
4348               72  Regional Manager         NaN
4484               75    Branch Manager         NaN
4659               79    Branch Manager         NaN
4930               87  District Manager         NaN
5103               89    Branch Manager         NaN
5325               96    Branch Manager  2022-09-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [ ]:
product_merge = pd.merge(order_details,product,left_on='PRODUCT_NUMBER',how='left',right_on='PRODUCT_NUMBER')
product_type_merge = pd.merge(product_merge, product_type, left_on='PRODUCT_TYPE_CODE',how='left',right_on='PRODUCT_TYPE_CODE')
product_not_sold = product_type_merge['QUANTITY'] > 750
product_type_merge.loc[product_not_sold ,['PRODUCT_NAME','PRODUCT_TYPE_EN']].drop_duplicates()

PRODUCT_NAME    PRODUCT_TYPE_EN
5417          Sun Shelter 30          Sunscreen
6801         BugShield Spray  Insect Repellents
6805   BugShield Lotion Lite  Insect Repellents
13909              Firefly 2           Lanterns
13985       BugShield Lotion  Insect Repellents
13988      BugShield Extreme  Insect Repellents
13995         Sun Shelter 15          Sunscreen
14000             Sun Shield          Sunscreen
16273               Star Peg              Tents

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [ ]:
product_merge_40_korting = ((product_merge['UNIT_PRICE'].astype(float) - product_merge['UNIT_SALE_PRICE'].astype(float))/product_merge['UNIT_PRICE'].astype(float)) > 0.4
product_merge.loc[product_merge_40_korting,['PRODUCT_NAME']].drop_duplicates()

PRODUCT_NAME
5594          BugShield Lotion Lite
6826              BugShield Extreme
9413              BugShield Natural
12276                Sun Shelter 30
14200              BugShield Lotion
14985               BugShield Spray
15124  Hailstorm Titanium Woods Set
20108                Sun Shelter 15

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [ ]:
returned_item_merge = pd.merge(returned_item,order_details,left_on='ORDER_DETAIL_CODE',how='left',right_on='ORDER_DETAIL_CODE')
returned_item_90 = (returned_item_merge['RETURN_QUANTITY'].astype(float)/returned_item_merge['QUANTITY'].astype(float)) < 0.9
returned_item_merge.loc[returned_item_90,['RETURN_QUANTITY','QUANTITY','RETURN_REASON_CODE']]

RETURN_QUANTITY  QUANTITY RETURN_REASON_CODE
1                 2        22                  2
4                 2        28                  1
5                 2        34                  1
6                 2        28                  2
8                 4        50                  2
..              ...       ...                ...
697              10       122                  2
698               6        70                  1
699               4        82                  1
700              12       128                  1
704               2        20                  1

[451 rows x 3 columns]

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [ ]:
product.groupby('PRODUCT_TYPE_CODE',as_index=False)['PRODUCT_NUMBER'].count()

PRODUCT_TYPE_CODE  PRODUCT_NUMBER
0                  1              10
1                 10               5
2                 11               5
3                 12               5
4                 13               4
5                 14               4
6                 15               5
7                 16               5
8                 17               5
9                 18               4
10                19               4
11                 2               6
12                20               3
13                21               4
14                 3               7
15                 4               6
16                 5              12
17                 6               4
18                 7               4
19                 8               7
20                 9               6

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [ ]:
retailer_site.groupby('COUNTRY_CODE',as_index=False)['RETAILER_SITE_CODE'].count()

COUNTRY_CODE  RETAILER_SITE_CODE
0             1                  29
1            10                  12
2            11                  25
3            12                  10
4            13                  10
5            14                  10
6            16                  10
7            17                  10
8            18                   6
9            19                   9
10            2                  32
11           20                  10
12           21                   1
13           22                   5
14            3                 101
15            4                  31
16            5                  10
17            6                  12
18            7                  20
19            8                  13
20            9                  25

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [38]:
product_merge = pd.merge(order_details,product,left_on='PRODUCT_NUMBER',how='left',right_on='PRODUCT_NUMBER')
product_merge_group = product_merge.groupby('PRODUCT_NAME',as_index=False)['QUANTITY'].sum()
Product_is_Cooking_Gear = product_merge_group['PRODUCT_TYPE_CODE'].astype(int) == 1
product_merge_group.loc[Product_is_Cooking_Gear,['PRODUCT_NAME','QUANTITY']]

KeyError: 'PRODUCT_TYPE_CODE'

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [1]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [2]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 